File: app/src/main/java/tv/withaibuild/customiuizer/mods/utils/HookerClassHelper.java
package tv.withaibuild.customiuizer.mods.utils;

import java.lang.reflect.Member;
import java.lang.reflect.Method;
import java.util.List;

import io.github.libxposed.api.XposedInterface;

/**
 * API 101 compatibility adapter for the module's existing before/after hooks.
 *
 * <p>The adapter deliberately keeps the API 100 callback behavior used by the Android 14
 * codebase, including mutable arguments, early return/throw and after-hook exception recovery.</p>
 */
public class HookerClassHelper {
    interface BeforeMethodCallback {
        void beforeHook(BeforeHookCallback callback);
    }

    interface AfterMethodCallback {
        void afterHook(AfterHookCallback callback);
    }

    public static class BeforeHookCallback {
        private static final Object[] EMPTY_ARGS = new Object[0];

        private final XposedInterface.Chain chain;
        private Member member;
        private Object thisObject;
        private Object[] args;
        private boolean thisObjectResolved;
        private boolean skipped;
        private Object result;
        private Throwable throwable;

        BeforeHookCallback(XposedInterface.Chain chain) {
            this.chain = chain;
        }

        public Member getMember() {
            if (member == null) member = chain.getExecutable();
            return member;
        }

        public Object getThisObject() {
            if (!thisObjectResolved) {
                thisObject = chain.getThisObject();
                thisObjectResolved = true;
            }
            return thisObject;
        }

        public Object[] getArgs() {
            if (args == null) {
                List<Object> argList = chain.getArgs();
                args = argList.isEmpty() ? EMPTY_ARGS : argList.toArray();
            }
            return args;
        }

        private boolean hasMaterializedArgs() {
            return args != null;
        }

        public void returnAndSkip(Object returnValue) {
            skipped = true;
            result = returnValue;
            throwable = null;
        }

        public void throwAndSkip(Throwable throwable) {
            skipped = true;
            result = null;
            this.throwable = throwable;
        }
    }

    public static class AfterHookCallback {
        private final BeforeHookCallback before;
        private final boolean skipped;
        private Object result;
        private Throwable throwable;

        AfterHookCallback(BeforeHookCallback before, Object result, Throwable throwable) {
            this.before = before;
            skipped = before.skipped;
            this.result = result;
            this.throwable = throwable;
        }

        public Member getMember() {
            return before.getMember();
        }

        public Object getThisObject() {
            return before.getThisObject();
        }

        public Object[] getArgs() {
            return before.getArgs();
        }

        public Object getResult() {
            return result;
        }

        public Throwable getThrowable() {
            return throwable;
        }

        public boolean isSkipped() {
            return skipped;
        }

        public void setResult(Object result) {
            this.result = result;
            throwable = null;
        }

        public void setThrowable(Throwable throwable) {
            result = null;
            this.throwable = throwable;
        }
    }

    public static class MethodHook implements BeforeMethodCallback, AfterMethodCallback, XposedInterface.Hooker {
        public final int mPriority;
        boolean mIsReturnConstant;
        Object mReturnConstantValue;
        private volatile byte afterCallbackState;

        public MethodHook() {
            this(XposedInterface.PRIORITY_DEFAULT);
        }

        public MethodHook(int priority) {
            mPriority = priority;
        }

        /**
         * Detects an after callback by signature instead of its source name.
         *
         * <p>R8 may rename {@code after(...)} in release builds. Looking it up with the literal
         * name "after" therefore disables every after hook after obfuscation, even though the
         * override is still present. The callback parameter type remains stable inside the same
         * optimized DEX and is safe to use for detection.</p>
         */
        private static boolean declaresAfterCallback(Class<?> hookClass) {
            Class<?> current = hookClass;
            while (current != null && current != MethodHook.class) {
                for (Method method : current.getDeclaredMethods()) {
                    Class<?>[] parameterTypes = method.getParameterTypes();
                    if (method.getReturnType() == Void.TYPE
                        && parameterTypes.length == 1
                        && parameterTypes[0] == AfterHookCallback.class) {
                        return true;
                    }
                }
                current = current.getSuperclass();
            }
            return false;
        }

        private boolean hasAfterCallback() {
            byte state = afterCallbackState;
            if (state == 0) {
                state = (byte) (declaresAfterCallback(getClass()) ? 2 : 1);
                afterCallbackState = state;
            }
            return state == 2;
        }

        @Override
        public Object intercept(XposedInterface.Chain chain) throws Throwable {
            if (mIsReturnConstant) {
                return mReturnConstantValue;
            }

            BeforeHookCallback before = new BeforeHookCallback(chain);
            beforeHook(before);

            Object result = before.result;
            Throwable throwable = before.throwable;
            if (!before.skipped) {
                try {
                    result = before.hasMaterializedArgs()
                        ? chain.proceed(before.args)
                        : chain.proceed();
                } catch (Throwable t) {
                    throwable = t;
                }
            }

            if (hasAfterCallback()) {
                AfterHookCallback after = new AfterHookCallback(before, result, throwable);
                afterHook(after);
                if (after.throwable != null) {
                    throw after.throwable;
                }
                return after.result;
            }

            if (throwable != null) {
                throw throwable;
            }
            return result;
        }

        public final void beforeHook(BeforeHookCallback callback) {
            try {
                before(callback);
            } catch (Throwable t) {
                XposedHelpers.log(t);
            }
        }

        public final void afterHook(AfterHookCallback callback) {
            try {
                after(callback);
            } catch (Throwable t) {
                XposedHelpers.log(t);
            }
        }

        protected void before(BeforeHookCallback callback) throws Throwable {
        }

        protected void after(AfterHookCallback callback) throws Throwable {
        }
    }

    public interface CustomMethodUnhooker {
        void unhook();
    }

    /** Skips the hooked method and returns {@code null}. */
    public static final MethodHook DO_NOTHING = new MethodHook(XposedInterface.PRIORITY_HIGHEST) {{
        mIsReturnConstant = true;
        mReturnConstantValue = null;
    }};

    /** Creates a highest-priority callback which always returns the supplied value. */
    public static MethodHook returnConstant(final Object result) {
        return returnConstant(XposedInterface.PRIORITY_HIGHEST, result);
    }

    /** Creates a callback which always returns the supplied value at the requested priority. */
    public static MethodHook returnConstant(int priority, final Object result) {
        return new MethodHook(priority) {{
            mIsReturnConstant = true;
            mReturnConstantValue = result;
        }};
    }
}

File: app/src/main/java/tv/withaibuild/customiuizer/mods/utils/HookerClassHelper.kt
package tv.withaibuild.customiuizer.mods.utils

import io.github.libxposed.api.XposedInterface
import java.lang.reflect.Member
import java.lang.reflect.Method

/**
 * API 101 compatibility adapter for the module's existing before/after hooks.
 *
 * The adapter deliberately keeps the API 100 callback behavior used by the Android 14
 * codebase, including mutable arguments, early return/throw and after-hook exception recovery.
 */
class HookerClassHelper private constructor() {

    interface BeforeMethodCallback {
        fun beforeHook(callback: BeforeHookCallback)
    }

    interface AfterMethodCallback {
        fun afterHook(callback: AfterHookCallback)
    }

    class BeforeHookCallback internal constructor(private val chain: XposedInterface.Chain) {
        private val EMPTY_ARGS = emptyArray<Any?>()

        private var member: Member? = null
        private var thisObject: Any? = null
        private var args: Array<Any?>? = null
        private var thisObjectResolved = false

        internal var skipped = false
        internal var result: Any? = null
        internal var throwable: Throwable? = null

        fun getMember(): Member {
            if (member == null) member = chain.executable
            return member!!
        }

        fun getThisObject(): Any? {
            if (!thisObjectResolved) {
                thisObject = chain.thisObject
                thisObjectResolved = true
            }
            return thisObject
        }

        fun getArgs(): Array<Any?> {
            if (args == null) {
                val argList: List<Any?> = chain.getArgs()
                args = if (argList.isEmpty()) EMPTY_ARGS else argList.toTypedArray()
            }
            return args!!
        }

        internal fun hasMaterializedArgs(): Boolean = args != null

        fun returnAndSkip(returnValue: Any?) {
            skipped = true
            result = returnValue
            throwable = null
        }

        fun throwAndSkip(throwable: Throwable) {
            skipped = true
            result = null
            this.throwable = throwable
        }
    }

    class AfterHookCallback internal constructor(
        private val before: BeforeHookCallback,
        result: Any?,
        throwable: Throwable?
    ) {
        val isSkipped: Boolean = before.skipped

        private var _result: Any? = result
        private var _throwable: Throwable? = throwable

        fun getMember(): Member = before.getMember()
        fun getThisObject(): Any? = before.getThisObject()
        fun getArgs(): Array<Any?> = before.getArgs()

        fun getResult(): Any? = _result
        fun setResult(result: Any?) {
            _result = result
            _throwable = null
        }

        fun getThrowable(): Throwable? = _throwable
        fun setThrowable(throwable: Throwable) {
            _result = null
            _throwable = throwable
        }
    }

    open class MethodHook : BeforeMethodCallback, AfterMethodCallback, XposedInterface.Hooker {
        @JvmField
        val mPriority: Int

        @JvmField
        var mIsReturnConstant = false

        @JvmField
        var mReturnConstantValue: Any? = null

        @Volatile
        private var afterCallbackState: Byte = 0

        constructor() : this(XposedInterface.PRIORITY_DEFAULT)

        constructor(priority: Int) {
            mPriority = priority
        }

        /**
         * Detects an after callback by signature instead of its source name.
         *
         * R8 may rename `after(...)` in release builds. Looking it up with the literal
         * name "after" therefore disables every after hook after obfuscation, even though the
         * override is still present. The callback parameter type remains stable inside the same
         * optimized DEX and is safe to use for detection.
         */
        private fun declaresAfterCallback(hookClass: Class<*>): Boolean {
            var current: Class<*>? = hookClass
            while (current != null && current != MethodHook::class.java) {
                for (method in current.declaredMethods) {
                    val parameterTypes = method.parameterTypes
                    if (method.returnType == Void.TYPE
                        && parameterTypes.size == 1
                        && parameterTypes[0] == AfterHookCallback::class.java
                    ) {
                        return true
                    }
                }
                current = current.superclass
            }
            return false
        }

        private fun hasAfterCallback(): Boolean {
            val state = afterCallbackState
            if (state.toInt() == 0) {
                val newState = if (declaresAfterCallback(javaClass)) 2 else 1
                afterCallbackState = newState.toByte()
            }
            return afterCallbackState.toInt() == 2
        }

        @Throws(Throwable::class)
        open override fun intercept(chain: XposedInterface.Chain): Any? {
            if (mIsReturnConstant) {
                return mReturnConstantValue
            }

            val before = BeforeHookCallback(chain)
            beforeHook(before)

            var result: Any? = before.result
            var throwable: Throwable? = before.throwable
            if (!before.skipped) {
                try {
                    result = if (before.hasMaterializedArgs())
                        chain.proceed(before.getArgs())
                    else
                        chain.proceed()
                } catch (t: Throwable) {
                    throwable = t
                }
            }

            if (hasAfterCallback()) {
                val after = AfterHookCallback(before, result, throwable)
                afterHook(after)
                if (after.getThrowable() != null) {
                    throw after.getThrowable()!!
                }
                return after.getResult()
            }

            if (throwable != null) {
                throw throwable
            }
            return result
        }

        override fun beforeHook(callback: BeforeHookCallback) {
            try {
                before(callback)
            } catch (t: Throwable) {
                XposedHelpers.log(t)
            }
        }

        override fun afterHook(callback: AfterHookCallback) {
            try {
                after(callback)
            } catch (t: Throwable) {
                XposedHelpers.log(t)
            }
        }

        @Throws(Throwable::class)
        protected open fun before(callback: BeforeHookCallback) {
        }

        @Throws(Throwable::class)
        protected open fun after(callback: AfterHookCallback) {
        }
    }

    interface CustomMethodUnhooker {
        fun unhook()
    }

    companion object {
        /** Skips the hooked method and returns `null`.  */
        @JvmField
        val DO_NOTHING: MethodHook = object : MethodHook(XposedInterface.PRIORITY_HIGHEST) {
            init {
                mIsReturnConstant = true
                mReturnConstantValue = null
            }
        }

        /** Creates a highest-priority callback which always returns the supplied value.  */
        @JvmStatic
        fun returnConstant(result: Any?): MethodHook = returnConstant(XposedInterface.PRIORITY_HIGHEST, result)

        /** Creates a callback which always returns the supplied value at the requested priority.  */
        @JvmStatic
        fun returnConstant(priority: Int, result: Any?): MethodHook {
            return object : MethodHook(priority) {
                init {
                    mIsReturnConstant = true
                    mReturnConstantValue = result
                }
            }
        }
    }
}
